common/cosmwasm-smart-contracts/service-provider-directory/src/types.rs

```diff
@@ -0,0 +1,63 @@
+use cosmwasm_std::{Addr, Coin};
+use serde::{Deserialize, Serialize};
+
+/// The directory of services are indexed by [`ServiceId`].
+pub type ServiceId = u32;
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
+pub struct Service {
+    /// The address of the service.
+    pub nym_address: NymAddress,
+    /// The service type.
+    pub service_type: ServiceType,
+    /// Service owner.
+    pub owner: Addr,
+    /// Block height at which the service was added.
+    pub block_height: u64,
+    /// The deposit used to announce the service.
+    pub deposit: Coin,
+}
+
+/// The types of addresses supported.
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, Debug)]
+pub enum NymAddress {
+    /// String representation of a nym address, which is of the form
+    /// client_id.client_enc@gateway_id.
+    Address(String),
+    // For the future when we have a nym-dns contract
+    //Name(String),
+}
+
+impl NymAddress {
+    /// Create a new nym address.
+    pub fn new(address: &str) -> Self {
+        Self::Address(address.to_string())
+    }
+
+    pub fn as_str(&self) -> &str {
+        match self {
+            NymAddress::Address(address) => address,
+        }
+    }
+}
+
+impl ToString for NymAddress {
```

Comment:
From what I remember, unless there's a very good reason to, it's discouraged to manually implement `ToString`. A preferred method is to implement `Display` and get `ToString` for free